### ISSUE TITLE
Fix `property-layout-mappings` false negatives for physical shorthand properties

### ISSUE THREAD

**Issue by mindplay-dk** [stylelint/stylelint#9420]
### What is the problem you're trying to solve?

I was surprised to find that shorthand margin/padding/border properties aren't being picked up - is there any reason these can't be converted to logical properties?

e.g. `margin: <block> <inline>` likely should be converted to `margin-block` and `margin-inline`, I think?

the 3- and 4-value properties could similarly be expanded and converted, perhaps conditional on their values.

Single-value properties of course are symmetrical and don't need any special handling - this works as it is.

But the other shorthands can be converted, I think.

This feature is likely being adopted by teams who designed for left-to-right languages and are now trying to migrate their code to support right-to-left languages, I think?

Here is a [motivating example](https://stylelint.io/demo/#N4Igxg9gJgpiBcIB0ALGAbdEAEwA6AdttgLYCGATgOYCWB82AjDCdgEwvYDMnALCwG5CxAC4UyBAM40RNCPVKVaBALToYAMxHYADNhhlJMIQQC+IADQgNNdQDkyJOIhgAPRwAd1SMJMmXweRsqBBA3DwgKbVgNMgBXdG18Imw8EHQJKjiyKhgAeQ9ZeUk0hmTiYjSoGgoYMCKCMnQZAE9S3GEKypAAIywwAGt2tJEIDxVRlR6IEVGSNItOrrS6ZoIYYfTNEQmIFQoaKhQRNKXsU07TRZSKBJhJMrO0jwoxmCiWtTIWiDid8g8HjoVBKCFS1iwAHd9hgyLIAG4bECXQimEwBSAEYIAMUi5BEoUkIha6jWIh8QUOSBIACt-FZYB5-IhyuCiSSMHQTmC0hkRPdudduuzSVyVJjgioiRIoJQoJs+QLTmZCCBTEA):

```css
.hello {
  margin: 1em 2em 3em 4em;
  transition: margin-left 0 ease;
}
```

This gets auto-fixed to:

```css
.hello {
  margin: 1em 2em 3em 4em;
  transition: margin-inline-start 0 ease;
}
```

So the transition was applied to `margin-left` which was the `2em` side - but it now being applied conditionally to either the left or right (`2em` or `4em`) side, which is probably going to give surprising results.

### What solution would you like to see?

To illustrate what was most likely the intended result, let's expand the shorthand properties:

```css
.hello {
  margin-top: 1em;
  margin-right: 2em;
  margin-bottom: 3em;
  margin-left: 4em;
  transition: margin-left 0 ease;
}
```

Feed [that](https://stylelint.io/demo/#N4Igxg9gJgpiBcIB0ALGAbdEAEwA6AdttgLYCGATgOYCWBAtAC4QAO82AjDCQNyHHlqdehRpUUjdgCZufIqUq0GAIwiNmJdgGZZ-BUIboYAM0nYALLvmMKZAgGcajGhALtBS+kdPYADNhgyexg5AF8QABoQYxojADkyEjhEGAAPRJYjJDB7e0jwVxiqBBA0lggKRmxYYzIAV3Qq-Hk8EHQ7KjqyKhgAeRZnV3tW9mbiYlaoGgoYMEGCMnQnAE8R3D1x7FblLDAAazXW5hYmCHpVdQgSVoiN8da6JYIYQ7aTRlORMQlWu9C9UK3eQUBoweyjO6tFgUVgwSrLLxkZYQOofcgsFh0KjDBBbaJYADuIgwZGcADcXiAAYRQnJ8pACEUAGIVciMEr2RjLIxPRjZQpiJAkABWeSisBYeUQYzxnO5GDojFe7UYYKVkT0rTlPMV9AZRXonLsUEoUGVpLVvwI-wIIFCQA) into the linter, and this auto-fixes to:

```css
.hello {
  margin-block-start: 1em;
  margin-inline-end: 2em;
  margin-block-end: 3em;
  margin-inline-start: 4em;
  transition: margin-inline-start 0 ease;
}
```

The transition is now being consistently applied to the `2em` side, in either writing direction, which is probably what you'd expect, right?

And we would need a similar treatment for 2- and 3-value shorthand properties.

Now, I realize we could ban shorthand properties and migrate all of those by hand first with a bunch of regex search and replace operations.

If that's the recommended solution, I would suggest this should be documented instead, as it's not going to be obvious to everyone, and since the default when enabling this rule could be a lot of broken CSS, and potentially things that would be *very* difficult to sort out later, after you realize what you've done. 😅

Personally, I would recommend and prefer something that "works as expected" - given that the shorthand properties are literally just shorthand (two syntaxes meaning the same thing) and given that the long form properties work as you'd expect, I think it would be natural to expect these to be handled consistently.

This may not *be* a bug (and I'm filing it as a feature request) but it sure *feels* like a bug. 😊

**Comment by mindplay-dk** [comment]
To others looking for a workaround, I am currently banning shorthand properties with more than 2 values, using this rule:

```
'declaration-property-max-values': {
  '/^margin$/': 2,
  '/^padding$/': 2,
  '/^border-width$/': 2,
  '/^border-style$/': 2,
  '/^border-color$/': 2,
},
```

I will manually resolve all of these by expanding offending properties to longer form syntax.

Note that this should be done *before* enabling the layout mappings rules, to ensure that those rules can then pick up and auto-fix the long form properties.

**Comment by jeddy3** [comment]
@mindplay-dk Thank you for the thorough report. It does feel like a bug, especially the example you gave of the inconsistent `transition` behaviour introduced by autofixing.

I think it was missed because none of the [prior art plugins](https://stylelint.io/user-guide/rules/property-layout-mappings/) that the rule is based on handle shorthands. I assume because the [`logical` keyword aspect of the spec](https://drafts.csswg.org/css-logical-1/#logical-shorthand-keyword) is in flux:

```css
.hello {
  margin: logical 1em 2em 3em 4em;
}
```

In the meantime, we could expand the following physical shorthand properties into their corresponding logical longhand ones:
- `inset`
- `margin`
- `padding`
- `border-width`
- `border-style`
- `border-color`
- `scroll-padding`
- `scroll-margin`

I think we'll need to be careful of things like:

```css
.hello {
  margin: 1em;
  transition: margin 0s ease;
}
```

Where `margin` in the `transition` property does mean "all" margins regardless of whether they're physical or logical. We'll need to dive into the specs to confirm this, though.

Let's also avoid expanding autofixing properties that contain `var()`, as the variable could hold any number of values.

I suspect other edge cases may present themselves during the implementation, but starting on the implementation is typically how we expose them.

I've labelled the issue as ready to implement. Please consider [contributing](https://stylelint.io/contributing) if you, or anyone else interested in fixing this bug, has time.

There are [steps on how to fix a bug in a rule](https://stylelint.io/contributor-guide/rules#fix-a-bug-in-a-rule) in the Contributor guide.

**Comment by ybiquitous** [comment]
> In the meantime, we could expand the following physical shorthand properties into their corresponding logical longhand ones:

@jeddy3 Just to confirm, does this not conflict with [`declaration-block-no-redundant-longhand-properties`](https://stylelint.io/user-guide/rules/declaration-block-no-redundant-longhand-properties)?

**Comment by jeddy3** [comment]
Whoops, you're correct! We should expand to the logical _shorthand_ ones.

For example: `margin` to `margin-block` and `margin-inline`.

Those won't conflict with `declaration-block-no-redundant-longhand-properties`. [Demo](https://stylelint.io/demo/#N4Igxg9gJgpiBcICGACYAdAdilBbJATgOYCWmAtAEYA2EYA1vCgIwAOAHgNxY77Fnky1MjCZsuWAL5YsqDNjyFSFAC4RWYjtwV9lVCCrW5NEnUoHUYAMxUntvcxQIkiAC1sstUmZjk9F-BQ0dPTkAM4qhB7MMLj2AXpCIuQwmFBMAEyx8boCwQwpaUwAzNn+uRRJmDDhkQQeACxlmJIgADQgViSWAHJIuHCIMOz9rJYAdGBhYe3gEJhdRAggw6wQ9SiwVkgArtQqaP7DKqlQYUwA2gDkEQCelsKYKuSQCy61SGmEUFcAum1SbSzV6LABi63wKmWdweZBUk3mi3GuAAVjMOrBWDNEPIcOgQDCYI8VPimPjqEgThF8QCFPjCcSXoj3hFPlBvqSUOTKTBqSBvJgQJIgA)

**Comment by ybiquitous** [comment]
> Whoops, you're correct! We should expand to the logical _shorthand_ ones.

Okay, that makes sense. 👍🏼
